Use nproc-2 to set parallelism for builds and tests

bthomee opened the pull request:
## High Level Overview of Change

This change reduces the number of cores used to build and test by two in CI.

### Context of Change

Using all cores may be contributing to occasional build and test failures. Only the Conan profile currently specifies that the number of cores minus one should be used for the level of parallelism, but that does not affect the building and testing as those are currently set to `$(nproc)`.

A new action was created in https://github.com/XRPLF/actions/pull/17, which this PR leverages to determine the number of processors supported by the OS on which the job is being executed, while subtracting 2 processors to reserve those for infrastructure-related tasks (e.g. some runners run on K8s). We also apply this to the Conan profile for consistency.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [X] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

legleux commented on .github/workflows/publish-docs.yml:
```diff
@@ -51,7 +51,7 @@ jobs:
           cd "${BUILD_DIR}"
           cmake -Donly_docs=ON ..
-          cmake --build . --target docs --parallel $(nproc)
+          cmake --build . --target docs --parallel $(( $(nproc) - 1 ))
```
You can let `nproc` do the math.
Probably use 2 to be safe.
`nproc --ignore 2`

bthomee replied:
I love it!

legleux commented:
> Although it would be nice to determine the number of cores to use in a single location, and then pass that value to all places that need it, it is very unwieldy to do so. 

https://github.com/XRPLF/clio/blob/develop/.github/actions/get-threads-number/action.yml

bthomee commented:
> > Although it would be nice to determine the number of cores to use in a single location, and then pass that value to all places that need it, it is very unwieldy to do so.
> 
> https://github.com/XRPLF/clio/blob/develop/.github/actions/get-threads-number/action.yml

I like that too, although it doesn't account for Windows. In our case all runners have nproc installed. I can create something similar.

bthomee commented:
> > > Although it would be nice to determine the number of cores to use in a single location, and then pass that value to all places that need it, it is very unwieldy to do so.
> > 
> > 
> > https://github.com/XRPLF/clio/blob/develop/.github/actions/get-threads-number/action.yml
> 
> I like that too, although it doesn't account for Windows. In our case all runners have nproc installed. I can create something similar.

@legleux I created https://github.com/XRPLF/actions/pull/17.

legleux commented:
Windoooooooooows! 👊

I can't believe all the years I've spent messing around shell math with nproc this was there the whole time? All it needs is a short flag.

legleux commented on conan/global.conf:
```diff
@@ -3,4 +3,4 @@
 core.download:parallel={{ os.cpu_count() }}
 core.upload:parallel={{ os.cpu_count() }}
-tools.build:jobs={{ os.cpu_count() - 1 }}
+tools.build:jobs={{ os.cpu_count() }}
```
Why does this get a pass? IIRC building boost is one dependency that causes issues. It might been worse using ninja. 

We can roll with it like this though and see if the additional logging cheng has added points to any particular contention.

bthomee replied:
This change applies _only_ when building locally on your own machine.

You can look at `.github/actions/build-deps/action.yml` where it is overridden to be nproc-2, which thus only applies when building in CI. You can for instance see in the "checks" tab for one of the Linux pipelines that `tools.build:jobs=30`, even though the number of processors detected was `32`.

legleux replied:
> This change applies only when building locally on your own machine.

Nothing applies this nor do we say to in the instructions (and I wouldn't because I'm pretty sure it'd kill my machine)?

bthomee replied:
Ah, you're right - I thought our build instructions referenced the global.conf file, but they don't.

Let me then completely remove that line to avoid any lingering confusion. I think it's fine to leave:
```
core.download:parallel={{ os.cpu_count() }}
core.upload:parallel={{ os.cpu_count() }}
```
as failures have not been observed during those stages afaik.

bthomee committed:
ci: Use nproc-2 to set parallelism for builds and tests

This change reduces the number of cores used to build and test, as using all cores may be contributing to occasional build and test failures.
